tests/src/core/testqgsrulebasedrenderer.cpp

```diff
@@ -143,6 +143,42 @@ class TestQgsRuleBasedRenderer: public QObject
       delete clone;
     }
 
+    /**
+     * test_many_rules_expression_filter checks that with > 50 rules we have a binary tree (log2(N))
+     */
+    void test_many_rules_expression_filter()
+    {
+
+      QgsVectorLayer *layer = new QgsVectorLayer( QStringLiteral( "point?field=fld:int" ), QStringLiteral( "x" ), QStringLiteral( "memory" ) );
+      QgsRenderContext ctx; // dummy render context
+      ctx.expressionContext().setFields( layer->fields() );
+
+      std::function<QString( const int ruleCount )> makeFilter = [ & ]( const int rc ) -> QString
+      {
+
+        // prepare renderer
+        RRule *rootRule = new RRule( nullptr );
+        for ( int i = 0; i < rc; i++ )
+        {
+          rootRule->appendChild( new RRule( nullptr, 0, 0, QStringLiteral( "%1" ).arg( i ) ) );
+        }
+        QgsRuleBasedRenderer r( rootRule );
+        r.startRender( ctx, layer->fields() );
+        r.stopRender( ctx );
+        return r.filter();
+      };
+
+      QCOMPARE( makeFilter( 1 ), QString( "(0)" ) );
+      QCOMPARE( makeFilter( 2 ), QString( "(0) OR (1)" ) );
```

Comment:
I don't really understand this - isn't the renderer filter supposed to return a full expression, like "col = 1 or col = 2"?